### PR TITLE
vscode: preserve arguments field of launch.json

### PR DESCRIPTION
**Pull request by qinggniq**
Signed-off-by: qinggniq <livewithblank@gmail.com>

`./tools/vscode/generate_debug_config.py` always replaces arguments of the previous config, preserve the arguments field will make it more convenient when we want to debug. 

It will preserve old argument only when no argument is set at the command line, and the old argument is not empty

Commit Message: preserve arguments field of config in launch.json
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by qinggniq** [comment]
@phlax

**Review by phlax — COMMENTED** [review]
@qinggniq im not entirely clear what this script does as im not a vscode user

i tried running it but it seemed to start compiling envoy so i quit

from source review tho im wondering:
- whether it would be better to add an `--overwrite` flag that just creates a new config
- whether in non-overwrite mode we should preserve all existing config customizations

i guess im wondering what the situation is for where you would want to run the script again when you already have an existing config

**Comment by qinggniq** [comment]
The `tools/vscode/generate_debug_config.py` do the following things:
1. Call `bazel build -c debug {target} {target}.dwp` to build target binary and debug information.
2. Use `bazel info bazel-bin` to get the directory path which target will store, use it to fill `target` in config. Use `bazel info workspace` or `directory` in `complier_database.json['directory']` to fill `debugger_args/--directory` in config.
3. Create or overwrite the config in `.vscode/launch.json`, use target name as the key, a target will only have one config.

In most cases, we just want to regenerate the binary and debug information after modifying the code when calling this script. 

I think what you said is a good idea:
> - whether it would be better to add an --overwrite flag that just creates a new config
> - whether in non-overwrite mode we should preserve all existing config customizations

in non-overwrite mode, we could just preserve the existing config customizations except `target` and `debugger_args` because they are generated by the script in step 2, I think they should not be modified manually, the other fields in config might have possible to be modified(I only custom `arguments`😄).

**Comment by phlax** [comment]
thanks @qinggniq for explanation

so iiuc this pr only preserves the `arguments` (or `args`) field - the others would be overwritten

i guess my question then is whether there is any value in preserving the others and/or adding logic to handle an overwrite flag

if we go with preserving more vars then i would use the pattern of reading/update/write the json

File: tools/vscode/generate_debug_config.py [hunk 1]
```diff
@@ -89,20 +89,38 @@ def lldb_config(target, binary, workspace, execroot, arguments):
     }
 
 
-def add_to_launch_json(target, binary, workspace, execroot, arguments, debugger_type):
-    launch = get_launch_json(workspace)
+def get_old_config(config_name, configurations):
```
Comment (phlax) [inline tools/vscode/generate_debug_config.py:92]: i dont think we need any of this - why not just replace

https://github.com/envoyproxy/envoy/blob/cda376394e1b73dc337e0f40c3b5560d2c338321/tools/vscode/generate_debug_config.py#L100-L107

with something like

```python

 configurations = launch.get("configurations", []) 
 always_overwritten_fields = ["x", "y", "z"]
 for config in configurations: 
     if config.get("name", None) == new_config["name"]: 
         if overwrite:
             config.clear() 
             config.update(new_config) 
         else:
             for k in always_overwritten_fields:
                 config[k] = new_config[k]
         break 
 else: 
     configurations.append(new_config) 

```

**Review by phlax — CHANGES_REQUESTED** [review]
@qinggniq iiuc i think we can change this more simply...

File: tools/vscode/generate_debug_config.py (hunk 1 above)
Comment (qinggniq) [reply tools/vscode/generate_debug_config.py:92]: agree, maybe `always_preserve_fields` is better?
```python
 configurations = launch.get("configurations", []) 
 always_preserve_fields = ["args"]
 for config in configurations: 
     if config.get("name", None) == new_config["name"]: 
         if overwrite:
             config.clear() 
             config.update(new_config) 
         else:
             for k not in always_preserve_fields:
                 old_config[k] = new_config[k]
             print(f"old config exists, fields {always_preserve_fields} will be preserved, use --overwirte to recreate config")
         break 
 else: 
     configurations.append(new_config) 
```

Comment (phlax) [reply tools/vscode/generate_debug_config.py:92]: > maybe always_preserve_fields is better?

wfm

Comment (phlax) [reply tools/vscode/generate_debug_config.py:92]: hmm - actually not `preserve` - its not clear what is being preserved ie old/new config

how about `always_update_fields` ?

Comment (qinggniq) [reply tools/vscode/generate_debug_config.py:92]: hh, for old config is preserve, for new config is update.

Comment (phlax) [reply tools/vscode/generate_debug_config.py:92]: tbh i think my preference is towards `always_overwrite_fields` as it matches to `--overwrite`

File: tools/vscode/generate_debug_config.py [hunk 2]
```diff
@@ -89,19 +89,31 @@ def lldb_config(target, binary, workspace, execroot, arguments):
...
-            config.clear()
-            config.update(new_config)
+            if overwrite:
+                config.clear()
+                config.update(new_config)
+            else:
+                for k in always_overwritten_fields:
+                    config[k] = new_config[k]
+                print(
+                    f"old config exists, only {always_overwritten_fields} will be updated, use --overwirte to recreate config"
```
Comment (phlax) [inline tools/vscode/generate_debug_config.py:115]: typo on `--overwrite`

File: tools/vscode/generate_debug_config.py [hunk 3]
```diff
@@ -112,13 +124,18 @@ def add_to_launch_json(target, binary, workspace, execroot, arguments, debugger_
...
     parser = argparse.ArgumentParser(description='Build and generate launch config for VSCode')
-    parser.add_argument('--debugger', default="gdb")
-    parser.add_argument('--args', default='')
-    parser.add_argument('target')
+    parser.add_argument('--debugger', default="gdb", help="debugger type, one of [gdb, lldb]")
+    parser.add_argument('--args', default='', help="command line arguments if target binary")
+    parser.add_argument(
+        '--overwrite',
+        action="store_true",
+        help="recreate config in launch config rather than preserve old config")
```
Comment (phlax) [inline tools/vscode/generate_debug_config.py:132]: perhaps

> recreate config without preserving any existing config

**Review by phlax — CHANGES_REQUESTED** [review]
@qinggniq almost there - a couple of nits...

**Review by phlax — APPROVED** [review]
lgtm once ci agrees, thanks @qinggniq

**Pull request merged**